packages/babel-parser/src/tokenizer/index.js

```diff
@@ -668,16 +698,34 @@ export default class Tokenizer extends LocationParser {
         this.finishToken(tt.comma);
         return;
       case charCodes.leftSquareBracket:
-        ++this.state.pos;
-        this.finishToken(tt.bracketL);
+        if (
+          this.getPluginOption("recordAndTuple", "syntaxType") === "bar" &&
+          this.input.charCodeAt(this.state.pos + 1) === charCodes.verticalBar
+        ) {
+          // [|
+          this.finishToken(tt.bracketBarL);
+          this.state.pos += 2;
+        } else {
+          ++this.state.pos;
+          this.finishToken(tt.bracketL);
+        }
         return;
       case charCodes.rightSquareBracket:
         ++this.state.pos;
         this.finishToken(tt.bracketR);
         return;
       case charCodes.leftCurlyBrace:
-        ++this.state.pos;
-        this.finishToken(tt.braceL);
+        if (
+          this.getPluginOption("recordAndTuple", "syntaxType") === "bar" &&
```

Comment:
Instead of checking this in the `if` head, can you do this in the body?

```js
if (this.getPluginOption("recordAndTuple", "syntaxType") !== "bar") {
  this.unexpected(this.state.pos, "Record expressions starting with '{|' are only allowed when the 'syntaxType' option of the 'recordAndTuple' plugin is set to 'bar'");
}
```

It would give us nicer error messages. (Also in the other places in this file).

Reply:
great idea! I pushed a commit that does this, and added test cases for each of the possible error messages.